feat(core): add `createComponent` and `reflectComponentType` functions

AndrewKushnir opened the pull request:
See individual commits.

Resolves https://github.com/angular/angular/issues/45263 and https://github.com/angular/angular/issues/44926.

## PR Type
What kind of change does this PR introduce?

- [x] Feature

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

AndrewKushnir committed:
feat(core): add `createComponent` function

This commit introduces a new function that allows creating a `ComponentRef` instance based on provided Component and a set of options. The function can be used to cover a number of use-cases where the `ComponentFactory` symbol was used previously.

Closes #45263.

AndrewKushnir committed:
feat(core): add `reflectComponentType` function

This commit introduces a new function that allows creating a object which exposes a number of getters to retrieve information about a given component.

Closes #44926.

gund commented:
This is great replacement for component factories!
However the naming maybe a bit confusing, maybe something like `ComponentMetadata` would be more appropriate to capture the intent instead of a `ComponentMirror` as it's not really mirroring anything but providing an extra "meta" information.

AndrewKushnir commented:
@gund thanks for the feedback. We discussed this question with the team further and decided to rename the function to `reflectComponentType`.

pkozlowski-opensource commented on packages/core/src/render3/component_ref.ts:
```diff
@@ -260,7 +261,7 @@ export function injectComponentFactoryResolver(): viewEngine_ComponentFactoryRes
  *
  */
-export class ComponentRef<T> extends viewEngine_ComponentRef<T> {
+export class ComponentRef<T> extends AbstractComponentRef<T> {
```
yay, great move, thnx!

pkozlowski-opensource commented on packages/core/src/render3/component_ref.ts:
```diff
@@ -307,3 +308,164 @@ export class ComponentRef<T> extends viewEngine_ComponentRef<T> {
+export function createComponentRef<C>(component: Type<C>, options: {
+  host: Element,
+  environmentInjector: EnvironmentInjector,
+  elementInjector?: Injector,
```
nit / question:  my understanding is that we are mostly using the term "node injector" and not "element injector". Should we rename this to `nodeInjector` and use this naming consistently?

AndrewKushnir replied:
Discussed offline: we are following the public-facing name for the NodeInjector: https://angular.io/guide/hierarchical-dependency-injection#two-injector-hierarchies.

pkozlowski-opensource commented on packages/core/src/render3/component_ref.ts:
```diff
@@ -307,3 +308,164 @@ export class ComponentRef<T> extends viewEngine_ComponentRef<T> {
+ * @returns
+ */
+export function createComponentRef<C>(component: Type<C>, options: {
+  host: Element,
```
host -> hostElement?

pkozlowski-opensource replied:
I don't think we should make the `host` required (!) - it is optional in https://angular.io/api/core/ComponentFactory#create _and_ is more flexible (allows passing selector).

I've encountered many use-cases in libraries where I want to create a `ComponentRef` _without_ attaching it anywhere. In those cases asking people to provide a `host` element is too much friction.

pkozlowski-opensource replied:
Technically speaking you _could_ create a host node yourself (ex. a div), without attaching it to a document. But yeh, this would be different, non-equivalent API and and too much burden in practice. We are going to clarify this bit.

AndrewKushnir replied:
Discussed this question offline. The PR is now updated to make the host element optional.

pkozlowski-opensource commented on packages/core/src/render3/errors.ts:
```diff
@@ -35,6 +30,16 @@ export function assertStandaloneComponentType(type: Type<unknown>) {
 }
 
+/** Verifies whether a given type is a component */
+export function assertComponentDef(type: Type<unknown>) {
```
Good refactor, thnx!

pkozlowski-opensource reviewed (changes requested):
LGTM overall but I would like to discuss the API surface, see https://github.com/angular/angular/pull/46685/files#r916504416

atscott reviewed (approved):
LGTM assuming the change from `createComponentRef` -> `createComponent` is made, as discussed.

AndrewKushnir commented:
@pkozlowski-opensource @atscott thanks for the feedback! I've updated the PR to incorporate the feedback, could you please take another look when you get a chance? Thank you.

mary-poppins commented:
You can preview d3d7f5e at https://pr46685-d3d7f5e.ngbuilds.io/.
You can preview 274f4a3 at https://pr46685-274f4a3.ngbuilds.io/.
You can preview 0a20df1 at https://pr46685-0a20df1.ngbuilds.io/.
You can preview 8f9c8e1 at https://pr46685-8f9c8e1.ngbuilds.io/.

pkozlowski-opensource commented on packages/core/test/acceptance/component_spec.ts:
```diff
@@ -639,4 +640,321 @@ describe('component', () => {
     expect(targetEl.innerHTML).toContain('DynamicComponent Content');
   });
+
+  describe('createComponent', () => {
```
EXCELLENT tests! I love how you though through all the cases / possibilities. Awesome work @AndrewKushnir !

pkozlowski-opensource reviewed (approved):
LGTM

I've left a couple of comments / remarks but there is nothing blocking for this PR. If there would be one thing to take care of it would be `AbstractComponentRef` showing up in the public API.

Reviewed-for: public-api 
Reviewed-for: fw-core

AndrewKushnir committed:
refactor(core): move the code to avoid circular dependencies

This commit refactors the code to move some functions around to avoid circular dependencies in TS imports. The newly added functions are now located in the `packages/core/src/render3/component.ts` file (instead of `packages/core/src/render3/component_ref.ts`), which is a better place for them anyway.

AndrewKushnir commented on packages/core/src/render3/component_ref.ts:
```diff
@@ -307,3 +308,163 @@ export class ComponentRef<T> extends viewEngine_ComponentRef<T> {
+  hostElement?: Element, environmentInjector: EnvironmentInjector,
+  elementInjector?: Injector,
+  projectableNodes?: Node[][],
+}): AbstractComponentRef<C> {
```
Great catch! I've ended up moving the functions around a bit to be able to reference the right symbols without circular deps.

mary-poppins commented:
You can preview 5dd8cad at https://pr46685-5dd8cad.ngbuilds.io/.

atscott reviewed (approved):
reviewed-for: public-api

AndrewKushnir commented:
[Presubmit](http://test/OCL:460518077:BASE:460518118:1657652044009:15cbd098).

mary-poppins commented:
You can preview f1af445 at https://pr46685-f1af445.ngbuilds.io/.

alxhub commented on goldens/public-api/core/index.md:
```diff
@@ -233,6 +233,22 @@ export abstract class ComponentFactoryResolver {
 }
 
+// @public
+export interface ComponentMirror<C> {
```
Nit: we can (and should) use `ReadonlyArray`s here.

alxhub commented on packages/core/src/render3/component.ts:
```diff
@@ -6,236 +6,183 @@
  */
-export function whenRendered(component: any): Promise<null> {
-  return getRootContext(component).clean;
+export function reflectComponentType<C>(component: Type<C>): ComponentMirror<C> {
```
We should return `null` here if `component` is not a component, instead of failing an assert.

alxhub commented on packages/core/src/render3/component_ref.ts:
```diff
@@ -307,3 +314,214 @@ export class ComponentRef<T> extends viewEngine_ComponentRef<T> {
+
+
+/** Options that control how the component should be bootstrapped. */
+export interface CreateComponentOptions {
```
Is this interface used? I think it might be leftover from `renderComponent`.

AndrewKushnir replied:
Yeah, that's a leftover from `renderComponent`, I'm removing it (as well as some other unused code) in https://github.com/angular/angular/pull/46806.

AndrewKushnir commented:
@alxhub thanks for the review, the comments are now addressed. Could you please take a final look at the public API when you get a chance? Thank you.

mary-poppins commented:
You can preview 5a47792 at https://pr46685-5a47792.ngbuilds.io/.

AndrewKushnir commented:
[Presubmit #2](http://test/OCL:460518077:BASE:460819371:1657751818565:48d9ecfb).

alxhub commented on packages/core/src/render3/component.ts:
```diff
@@ -6,236 +6,184 @@
+  /**
+   * The inputs of the component.
+   */
+  get inputs(): ReadonlyArray<{propName: string, templateName: string}>;
```
In keeping with the `ReadonlyArray` theme, maybe make `propName` and `templateName` readonly here as well.

AndrewKushnir replied:
Good catch 👍 I've updated the PR.

alxhub reviewed (approved):
Reviewed-for: public-api

AndrewKushnir committed:
refactor(core): annotate `ComponentMirror` object fields as readonly

This commit updates the `inputs` and `outputs` objects of the `ComponentMirror` to mark their properties as readonly.

mary-poppins commented:
You can preview cb82edc at https://pr46685-cb82edc.ngbuilds.io/.

jessicajaniuk commented:
This PR was merged into the repository by commit 4a81fe7d32430785bad6d2b8334b68671c1687f1.
